### ISSUE TITLE
There seems to have a limit for how many combinations of f4 can be computed at once

### ISSUE THREAD

**Issue by dangliu** [bodkan/admixr#86]
Example of codes
```
# Groups
Massim <- c(Collingwood_Bay, W_Mas, N_Mas, S_Mas)
PNG_L <- c("Madang_Lowland", "Morobe", "East_Sepik", "Western", "Central")
Bismark <- c("Manus_New_Ireland", "West_New_Britain", "East_New_Britain")
Solomon <- c("Bougainville", "Vella_Lavella", "Malaita", "Santa_Cruz", "Bellona_Rennell", "Tikopia")

# f4 
f4_x <- f4(W = c(Massim, PNG_L,Bismark, Solomon), X = "Southern_Highlands", Y = "EastAsia_AN", Z = "Australia", data = data) # will fail
f4_x_1 <- f4(W = c(Massim, PNG_L), X = "Southern_Highlands", Y = "EastAsia_AN", Z = "Australia", data = data) # can run
f4_x_2 <- f4(W = c(Bismark, Solomon), X = "Southern_Highlands", Y = "EastAsia_AN", Z = "Australia", data = data) # can run
```

Error message when it fails:
```
Killed
BEGINNING OF OUTPUT FILE
==================================================

qpDstat: parameter file: /tmp/RtmpqS3A9E/qpDstat__486742862.par
### THE INPUT PARAMETERS
##PARAMETER NAME: VALUE
genotypename: /mnt/scratch/dang/Kula/f4/Kula.meta.filtered.geno
snpname: /mnt/scratch/dang/Kula/f4/Kula.meta.filtered.snp
indivname: /mnt/scratch/dang/Kula/f4/Kula.meta.filtered.ind
popfilename: /tmp/RtmpqS3A9E/qpDstat__486742862.pop
f4mode: YES
printsd: YES
## qpDstat version: 980
inbreed set NO
number of quadruples 29

==================================================
END OF OUTPUT FILE


Error: The output file we got from ADMIXTOOLS is truncated.
       Please examine the full output above and check for errors.
```

**Comment by bodkan** [comment]
Hello Dang,

Thanks again for submitting this as a GitHub issue.

The error message you posted makes things quite a bit clearer -- "Killed" here almost always means that ADMIXTOOLS ran out of memory. This is almost certain in this case because the output was truncated. This is _admixr_ telling us that it didn't receive a complete log file from ADMIXTOOLS.

Can you rerun this crashing analysis and monitor the memory usage somehow? Are you running it on one of "our" MPI number crunchers? I bet you could see the qpDstat process approaching max memory and being killed.

In other words, I don't think this is any change in ADMIXTOOLS. Perhaps you're running this analysis on a much larger data set than before?

A follow up question: what is the source of data you're using? How many SNPs? How many individuals per group? Analysis of this kind shouldn't get killed on a large enough computer (i.e. our number crunchers) for something like Human Origins etc. Are you running this on whole-genome shotgun data?

**Comment by dangliu** [comment]
Hi Martin,

I see, do you know how can I print the memory usage to a log file?

I am running it on a merged data of Oceanian populations with the Illumina MEGA array SNPs. The number of individuals in total is 776, and the number of SNPs is 1408767. The number of individuals per group is range from 1 to 41.

Thanks for answering,
Dang

**Comment by bodkan** [comment]
I'm afraid there's no option for realtime tracking of memory usage.

You could verify this by running `htop` in another terminal, monitoring the RAM usage as the analysis progresses.

Are you running this on your laptop? I'm basically 99.99% sure this is an out-of-memory problem. Unfortunately it's been a long time since I ran large-scale admixr analyses, so I'm out of touch w.r.t. what is the average memory consumption.

I remember for my PNAS paper on Neanderthal decline, I ran things on all available HO+1240k capture data and had to split populations in smaller chunks to avoid memory crashes.

**Comment by dangliu** [comment]
Good to know it's just the memory problem : D

I am running on bionc server. It was fine with running on the Thai HO data also ~700 individuals, but the MEGA array doubles the number of SNPs even after filtering, so probably that's why.

**Comment by bodkan** [comment]
Oh yeah, I bet this is the reason.

Well, as I said, I have ran into this problem routinely when I was doing f4 analyses with shotgun SGDP data (many more SNPs than on the arrays you're using). The way around that was to partition the individuals (the vectors of character ID names) into smaller pieces.

Because admixr always gives you a data frame, it's easy to split the analysis in smaller chunks, and then simply use `dplyr::bind_rows` to merge that in a single data frame.

I will now close the issue. It will be useful for directing people here who also run into this "Killed" issue.

Thanks for your help figuring it out!
